crates/ty_python_semantic/src/types/constraints.rs

```diff
@@ -273,1231 +239,1147 @@ impl<'db> ConstraintSet<'db> {
-            }
+    /// Returns a new BDD with any occurrence of `left ∧ right` replaced with `replacement`.
+    fn substitute_intersection(
+        self,
+        db: &'db dyn Db,
+        left: ConstraintAssignment<'db>,
+        right: ConstraintAssignment<'db>,
+        replacement: Node<'db>,
+    ) -> Self {
+        // If left and right are not both present in the input BDD, we should not even attempt the
+        // substitution, since the Shannon expansion might introduce the missing variables! That
+        // confuses us below when we try to detect whether the substitution is consistent with the
+        // input.
+        if !self.constraint_in_domain(db, left.constraint())
+            || !self.constraint_in_domain(db, right.constraint())
+        {
+            return self;
+        }
+
+        // We perform a Shannon expansion to find out what the input BDD evaluates to when:
+        //   - left and right are both true
+        //   - left is false
+        //   - left is true and right is false
+        // This covers the entire truth table of `left ∧ right`.
+        let when_left_and_right = self.restrict(db, [left, right]);
+        let when_not_left = self.restrict(db, [left.flipped()]);
+        let when_left_but_not_right = self.restrict(db, [left, right.flipped()]);
+
+        // The result should test `replacement`, and when it's true, it should produce the same
+        // output that input would when `left ∧ right` is true. When replacement is false, it
+        // should fall back on testing left and right individually to make sure we produce the
+        // correct outputs in the `¬(left ∧ right)` case. So the result is
+        //
+        //   if replacement
+        //     when_left_and_right
+        //   else if not left
+        //     when_not_left
+        //   else if not right
+        //     when_left_but_not_right
+        //   else
+        //     false
+        //
+        //  (Note that the `else` branch shouldn't be reachable, but we have to provide something!)
+        let left_node = Node::new_satisfied_constraint(db, left);
+        let right_node = Node::new_satisfied_constraint(db, right);
+        let right_result = right_node.ite(db, Node::AlwaysFalse, when_left_but_not_right);
+        let left_result = left_node.ite(db, right_result, when_not_left);
+        let result = replacement.ite(db, when_left_and_right, left_result);
+
+        // Lastly, verify that the result is consistent with the input. (It must produce the same
+        // results when `left ∧ right`.) If it doesn't, the substitution isn't valid, and we should
+        // return the original BDD unmodified.
+        let validity = replacement.iff(db, left_node.and(db, right_node));
+        let constrained_original = self.and(db, validity);
+        let constrained_replacement = result.and(db, validity);
+        if constrained_original == constrained_replacement {
+            result
+        } else {
+            self
```

Comment:
When would we attempt a simplification that this validation would reject? Are some of the simplifications attempted in `InteriorNode::simplify` speculative, or possibly invalid? If so, maybe we should mark those more clearly than we do now?